# Напишите заголовок проекта здесь

- Автор:
- Дата:

### Цели и задачи проекта

<font color='#777778'>Перечислите цель проекта и задачи, которые вы решаете. Для этого можно использовать описание проекта, но полезнее будет сформулировать основную цель проекта самостоятельно.</font>

### Описание данных

<font color='#777778'>Здесь приведите описание данных.</font>

### Содержимое проекта

<font color='#777778'>Перечислите основные шаги проекта или напишите оглавление. Используйте описание проекта, чтобы зафиксировать основные шаги.</font>

---

## 1. Загрузка данных и знакомство с ними

- Загрузите необходимые библиотеки Python и данные датасета `/datasets/new_games.csv`.


In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('/datasets/new_games.csv')
df.head()

Name Platform  Year of Release         Genre  NA sales  \
0                Wii Sports      Wii           2006.0        Sports     41.36   
1         Super Mario Bros.      NES           1985.0      Platform     29.08   
2            Mario Kart Wii      Wii           2008.0        Racing     15.68   
3         Wii Sports Resort      Wii           2009.0        Sports     15.61   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing     11.27   

  EU sales JP sales  Other sales  Critic Score User Score Rating  
0    28.96     3.77         8.45          76.0          8      E  
1     3.58     6.81         0.77           NaN        NaN    NaN  
2    12.76     3.79         3.29          82.0        8.3      E  
3    10.93     3.28         2.95          80.0          8      E  
4     8.89    10.22         1.00           NaN        NaN    NaN

- Познакомьтесь с данными: выведите первые строки и результат метода `info()`.


In [3]:
df.describe()

Year of Release      NA sales   Other sales  Critic Score
count     16681.000000  16956.000000  16956.000000   8242.000000
mean       2006.485522      0.262023      0.047087     68.926717
std           5.873102      0.808654      0.185577     13.944565
min        1980.000000      0.000000      0.000000     13.000000
25%        2003.000000      0.000000      0.000000     60.000000
50%        2007.000000      0.080000      0.010000     71.000000
75%        2010.000000      0.240000      0.030000     79.000000
max        2016.000000     41.360000     10.570000     98.000000

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


- Сделайте вывод о полученных данных: данные какого объёма вам предоставили, соответствуют ли они описанию, встречаются ли в них пропуски, используются ли верные типы данных.
- Отметьте другие особенности данных, которые вы обнаружили и на которые стоит обратить внимание при предобработке. Например, вы можете проверить названия столбцов: все ли названия отражают содержимое данных и прописаны в удобном для работы виде.

Комментарии по датафрейму:
- Существуют пропуски в столбцах:
  - Name
  - Year of Release
  - Genre
  - Critic Score: около половины значения NaN
  - User Score
  - Rating
- Типы данных:
  - Year of Release: сейчас тип float, но должен быть int
  - EU sales: сейчас тип object, но должен быть float
  - JP sales: сейчас тип object, но должен быть float
  - Critic Score: сейчас тип float, но можно сделать int
  - User Score: сейчас тип object, но можно убрать значения 'tbd', и преобразовать в float
- Названия столбцов не приведены к snake_case

---

## 2.  Проверка ошибок в данных и их предобработка


### 2.1. Названия, или метки, столбцов датафрейма

- Выведите на экран названия всех столбцов датафрейма и проверьте их стиль написания.
- Приведите все столбцы к стилю snake case. Названия должны быть в нижнем регистре, а вместо пробелов — подчёркивания.

In [5]:
df.columns

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')

In [6]:
# Преобразование названиq столбцов в snake_case
for column in df.columns:
    snake_column = column.lower().replace(' ', '_')
    df = df.rename(columns={column : snake_column})

In [7]:
# Проверка
df.head()

name platform  year_of_release         genre  na_sales  \
0                Wii Sports      Wii           2006.0        Sports     41.36   
1         Super Mario Bros.      NES           1985.0      Platform     29.08   
2            Mario Kart Wii      Wii           2008.0        Racing     15.68   
3         Wii Sports Resort      Wii           2009.0        Sports     15.61   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing     11.27   

  eu_sales jp_sales  other_sales  critic_score user_score rating  
0    28.96     3.77         8.45          76.0          8      E  
1     3.58     6.81         0.77           NaN        NaN    NaN  
2    12.76     3.79         3.29          82.0        8.3      E  
3    10.93     3.28         2.95          80.0          8      E  
4     8.89    10.22         1.00           NaN        NaN    NaN

### 2.2. Типы данных

- Если встречаются некорректные типы данных, предположите их причины.
- При необходимости проведите преобразование типов данных. Помните, что столбцы с числовыми данными и пропусками нельзя преобразовать к типу `int64`. Сначала вам понадобится обработать пропуски, а затем преобразовать типы данных.

In [8]:
for column in df.columns:
    print(f"Тип данных столбца {column}: {df[column].dtype}")

Тип данных столбца name: object
Тип данных столбца platform: object
Тип данных столбца year_of_release: float64
Тип данных столбца genre: object
Тип данных столбца na_sales: float64
Тип данных столбца eu_sales: object
Тип данных столбца jp_sales: object
Тип данных столбца other_sales: float64
Тип данных столбца critic_score: float64
Тип данных столбца user_score: object
Тип данных столбца rating: object


Комментарии:
  - Столбец year_of_release - тип float, должен быть int. Возможно есть какая-то ошибка с и где-то десятичная часть не равна 0, в любом случае округление до int решает проблему
  - Столбец eu_sales - тип object, должен быть float. Скорее всего есть строковые значения, которые невозможно преобразовать в числовые
  - Столбец jp_sales - тип object, должен быть float. Скорее всего есть строковые значения, которые невозможно преобразовать в числовые
  - Столбец user_score - тип object, должен быть float. Скорее всего есть строковые значения, которые невозможно преобразовать в числовые

- В числовых столбцах могут встретиться строковые значения, например `unknown` или другие. Приводите такие столбцы к числовому типу данных, заменив строковые значения на пропуски.

In [9]:
# Создаем копию датафрейма для будущего сравнения с начальным
c_df = df.copy()

In [10]:
# Пытаемся преобразовать столбец eu_sales к нужному типу, тем самым проверяя есть ли неподходящие значения
# c_df['eu_sales'] = c_df['eu_sales'].astype('float64') # нашли что в столбце существуют строковые 'unknown'

In [11]:
# Выводим строки с "unknown" в столбце eu_sales
print(f"Количество строк с 'unknown' = {len(c_df[c_df['eu_sales']=='unknown'])}")

# Преобразуем значения 'unknown' в NaN
c_df['eu_sales'] = pd.to_numeric(c_df['eu_sales'], errors='coerce')

print(f"Количество строк с 'unknown' = {len(c_df[c_df['eu_sales']=='unknown'])}")

Количество строк с 'unknown' = 6
Количество строк с 'unknown' = 0


In [12]:
# Пытаемся преобразовать столбец jp_sales к нужному типу, тем самым проверяя есть ли неподходящие значения
# c_df['jp_sales'] = c_df['jp_sales'].astype('float64') # нашли что в столбце существуют строковые 'unknown'

In [13]:
print(f"Количество строк с 'unknown' = {len(c_df[c_df['jp_sales']=='unknown'])}")

# Преобразуем значения 'unknown' в NaN
c_df['jp_sales'] = pd.to_numeric(c_df['jp_sales'], errors='coerce')

print(f"Количество строк с 'unknown' = {len(c_df[c_df['jp_sales']=='unknown'])}")

Количество строк с 'unknown' = 4
Количество строк с 'unknown' = 0


In [14]:
# Пытаемся преобразовать столбец user_score в float и заодно проверяем есть ли неподходящие значения
# c_df['user_score'] = c_df['user_score'].astype('float64') # нашли что в столбце есть строковое 'tbd'

In [15]:
print(f"Количество строк с 'tbd' в столбце user_score: {len(c_df[c_df['user_score']=='tbd'])}")

print(f"Минимальное значение в столбце user_score = {pd.to_numeric(c_df['user_score'], errors='coerce').min()}")

print(f"Максимальное значение в столбце user_score = {pd.to_numeric(c_df['user_score'], errors='coerce').max()}")

Количество строк с 'tbd' в столбце user_score: 2464
Минимальное значение в столбце user_score = 0.0
Максимальное значение в столбце user_score = 9.7


В столбце user_score есть значения 'tbd', значит для некоторых игр user_score еще уточняется.
Скорее всего user_score имеет имеет оценку от 0 до 10. Зная об этом, можем заменить 'tbd' на значение-индикатор -1

In [16]:
def tbd_replace(row):
    if row['user_score']=='tbd':
        return -1
    else:
        return row['user_score']

c_df['user_score'] = c_df.apply(tbd_replace, axis=1)

### 2.3. Наличие пропусков в данных

- Посчитайте количество пропусков в каждом столбце в абсолютных и относительных значениях.


In [17]:
# Количество пропусков в каждом столбце
for column in c_df.columns:
    print(f"Количество пустых значений в столбце {column}: {len(df) - c_df[column].count()}")
    print(f"Доля пустых значений в столбце {column}: {round((len(df) - c_df[column].count()) / len(df) * 100, 2)}%")
    print()

Количество пустых значений в столбце name: 2
Доля пустых значений в столбце name: 0.01%

Количество пустых значений в столбце platform: 0
Доля пустых значений в столбце platform: 0.0%

Количество пустых значений в столбце year_of_release: 275
Доля пустых значений в столбце year_of_release: 1.62%

Количество пустых значений в столбце genre: 2
Доля пустых значений в столбце genre: 0.01%

Количество пустых значений в столбце na_sales: 0
Доля пустых значений в столбце na_sales: 0.0%

Количество пустых значений в столбце eu_sales: 6
Доля пустых значений в столбце eu_sales: 0.04%

Количество пустых значений в столбце jp_sales: 4
Доля пустых значений в столбце jp_sales: 0.02%

Количество пустых значений в столбце other_sales: 0
Доля пустых значений в столбце other_sales: 0.0%

Количество пустых значений в столбце critic_score: 8714
Доля пустых значений в столбце critic_score: 51.39%

Количество пустых значений в столбце user_score: 6804
Доля пустых значений в столбце user_score: 40.13%

Колич

- Изучите данные с пропущенными значениями. Напишите промежуточный вывод: для каких столбцов характерны пропуски и сколько их. Предположите, почему пропуски могли возникнуть. Укажите, какие действия с этими данными можно сделать и почему.


In [18]:
# Строки с пустым 'name'
c_df[c_df['name'].isna()]

name platform  year_of_release genre  na_sales  eu_sales  jp_sales  \
661    NaN      GEN           1993.0   NaN      1.78      0.53      0.00   
14439  NaN      GEN           1993.0   NaN      0.00      0.00      0.03   

       other_sales  critic_score user_score rating  
661           0.08           NaN        NaN    NaN  
14439         0.00           NaN        NaN    NaN

In [19]:
# Строки с пустым year_of_release
c_df[c_df['year_of_release'].isna()].head()

name platform  year_of_release     genre  na_sales  \
183             Madden NFL 2004      PS2              NaN    Sports      4.26   
379            FIFA Soccer 2004      PS2              NaN    Sports      0.59   
458  LEGO Batman: The Videogame      Wii              NaN    Action      1.80   
477  wwe Smackdown vs. Raw 2006      PS2              NaN  Fighting      1.57   
611              Space Invaders     2600              NaN   Shooter      2.36   

     eu_sales  jp_sales  other_sales  critic_score user_score rating  
183      0.26      0.01         0.71          94.0        8.5      E  
379      2.36      0.04         0.51          84.0        6.4      E  
458      0.97      0.00         0.29          74.0        7.9   E10+  
477      1.02      0.00         0.41           NaN        NaN    NaN  
611      0.14      0.00         0.03           NaN        NaN    NaN

In [20]:
# Проверяем распределние пустых строк в 'year_of_release' по платформам
by_plat = c_df[c_df['year_of_release'].isna()]
print(by_plat['platform'].value_counts())

print()

# Проверяем распределние пустых строк в 'year_of_release' по жанрам
by_genre = c_df[c_df['year_of_release'].isna()]
by_genre['genre'].value_counts()

PS2     35
Wii     35
X360    32
DS      30
PS3     25
XB      21
PC      18
PSP     17
2600    17
GC      14
GBA     11
3DS      8
PS       7
N64      3
PSV      1
GB       1
Name: platform, dtype: int64



Action          64
Sports          42
Misc            29
Shooter         27
Racing          23
Role-Playing    17
Simulation      16
Fighting        12
Puzzle          11
Strategy        11
Adventure       10
Platform        10
ACTION           1
SPORTS           1
SIMULATION       1
Name: genre, dtype: int64

In [21]:
# Строки с пустым 'critic_score'
nan_critic_score = c_df[c_df['critic_score'].isna()]
nan_critic_score.head()

name platform  year_of_release         genre  \
1          Super Mario Bros.      NES           1985.0      Platform   
4   Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
5                     Tetris       GB           1989.0        Puzzle   
9                  Duck Hunt      NES           1984.0       Shooter   
10                Nintendogs       DS           2005.0    Simulation   

    na_sales  eu_sales  jp_sales  other_sales  critic_score user_score rating  
1      29.08      3.58      6.81         0.77           NaN        NaN    NaN  
4      11.27      8.89     10.22         1.00           NaN        NaN    NaN  
5      23.20      2.26      4.22         0.58           NaN        NaN    NaN  
9      26.93      0.63      0.28         0.47           NaN        NaN    NaN  
10      9.05     10.95      1.93         2.74           NaN        NaN    NaN

In [22]:
# Проверяем зависимость critic_score от года выпуска игры
nan_by_year = nan_critic_score.groupby(by='year_of_release')['year_of_release'].count().reset_index(name='nan_count')
all_by_year = c_df.groupby(by='year_of_release')['year_of_release'].count().reset_index(name='all_count')

merged = pd.merge(nan_by_year, all_by_year, on='year_of_release')
merged['share_nan_in_all'] = round(merged['nan_count'] / merged['all_count'], 2)
merged

year_of_release  nan_count  all_count  share_nan_in_all
0            1980.0          9          9              1.00
1            1981.0         46         46              1.00
2            1982.0         37         37              1.00
3            1983.0         18         18              1.00
4            1984.0         14         14              1.00
5            1985.0         13         14              0.93
6            1986.0         22         22              1.00
7            1987.0         17         17              1.00
8            1988.0         14         15              0.93
9            1989.0         17         17              1.00
10           1990.0         16         16              1.00
11           1991.0         42         42              1.00
12           1992.0         42         43              0.98
13           1993.0         62         62              1.00
14           1994.0        120        121              0.99
15           1995.0        220        220              1.00
16           1996.0        259        267              0.97
17           1997.0        276        293              0.94
18           1998.0        356        384              0.93
19           1999.0        301        341              0.88
20           2000.0        213        357              0.60
21           2001.0        158        491              0.32
22           2002.0        205        839              0.24
23           2003.0        194        789              0.25
24           2004.0        204        771              0.26
25           2005.0        288        950              0.30
26           2006.0        394       1020              0.39
27           2007.0        511       1218              0.42
28           2008.0        720       1445              0.50
29           2009.0        791       1450              0.55
30           2010.0        770       1279              0.60
31           2011.0        646       1149              0.56
32           2012.0        345        670              0.51
33           2013.0        274        552              0.50
34           2014.0        322        584              0.55
35           2015.0        385        612              0.63
36           2016.0        275        507              0.54

In [23]:
# Проверяем зависимость critic_score от платформы
nan_by_platform = nan_critic_score.groupby(by='platform')['platform'].count().reset_index(name='nan_count')
all_by_platform = c_df.groupby(by='platform')['platform'].count().reset_index(name='all_count')
merged = pd.merge(nan_by_platform, all_by_platform, on='platform')
merged['share_nan_in_all'] = round(merged['nan_count'] / merged['all_count'], 2)
merged

platform  nan_count  all_count  share_nan_in_all
0      2600        135        135              1.00
1       3DO          3          3              1.00
2       3DS        361        530              0.68
3        DC         38         52              0.73
4        DS       1453       2177              0.67
5        GB         98         98              1.00
6       GBA        394        837              0.47
7        GC        110        563              0.20
8       GEN         29         29              1.00
9        GG          1          1              1.00
10      N64        323        323              1.00
11      NES        100        100              1.00
12       NG         12         12              1.00
13       PC        264        990              0.27
14     PCFX          1          1              1.00
15       PS       1013       1215              0.83
16      PS2        875       2189              0.40
17      PS3        523       1355              0.39
18      PS4        143        395              0.36
19      PSP        760       1229              0.62
20      PSV        312        435              0.72
21      SAT        174        174              1.00
22      SCD          6          6              1.00
23     SNES        241        241              1.00
24     TG16          2          2              1.00
25       WS          6          6              1.00
26      Wii        749       1340              0.56
27     WiiU         57        147              0.39
28     X360        350       1281              0.27
29       XB        101        839              0.12
30     XOne         80        251              0.32

In [24]:
# Строки с пустым 'user_score'
nan_user_score = c_df[c_df['user_score'].isna()]
nan_user_score.head()

name platform  year_of_release         genre  \
1          Super Mario Bros.      NES           1985.0      Platform   
4   Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
5                     Tetris       GB           1989.0        Puzzle   
9                  Duck Hunt      NES           1984.0       Shooter   
10                Nintendogs       DS           2005.0    Simulation   

    na_sales  eu_sales  jp_sales  other_sales  critic_score user_score rating  
1      29.08      3.58      6.81         0.77           NaN        NaN    NaN  
4      11.27      8.89     10.22         1.00           NaN        NaN    NaN  
5      23.20      2.26      4.22         0.58           NaN        NaN    NaN  
9      26.93      0.63      0.28         0.47           NaN        NaN    NaN  
10      9.05     10.95      1.93         2.74           NaN        NaN    NaN

In [25]:
# Проверяем зависимость user_score от года выпуска игры
nan_by_year = nan_user_score.groupby(by='year_of_release')['year_of_release'].count().reset_index(name='nan_count')
all_by_year = c_df.groupby(by='year_of_release')['year_of_release'].count().reset_index(name='all_count')
merged = pd.merge(nan_by_year, all_by_year, on='year_of_release')
merged['share_nan_in_all'] = round(merged['nan_count'] / merged['all_count'], 2)
merged

year_of_release  nan_count  all_count  share_nan_in_all
0            1980.0          9          9              1.00
1            1981.0         46         46              1.00
2            1982.0         37         37              1.00
3            1983.0         18         18              1.00
4            1984.0         14         14              1.00
5            1985.0         13         14              0.93
6            1986.0         22         22              1.00
7            1987.0         17         17              1.00
8            1988.0         14         15              0.93
9            1989.0         17         17              1.00
10           1990.0         16         16              1.00
11           1991.0         42         42              1.00
12           1992.0         42         43              0.98
13           1993.0         62         62              1.00
14           1994.0        120        121              0.99
15           1995.0        220        220              1.00
16           1996.0        259        267              0.97
17           1997.0        274        293              0.94
18           1998.0        353        384              0.92
19           1999.0        301        341              0.88
20           2000.0        205        357              0.57
21           2001.0        145        491              0.30
22           2002.0        176        839              0.21
23           2003.0        167        789              0.21
24           2004.0        169        771              0.22
25           2005.0        237        950              0.25
26           2006.0        337       1020              0.33
27           2007.0        381       1218              0.31
28           2008.0        446       1445              0.31
29           2009.0        415       1450              0.29
30           2010.0        465       1279              0.36
31           2011.0        424       1149              0.37
32           2012.0        301        670              0.45
33           2013.0        228        552              0.41
34           2014.0        233        584              0.40
35           2015.0        274        612              0.45
36           2016.0        208        507              0.41

In [26]:
# Проверяем зависимость user_score от платформы игры
nan_by_platform = nan_user_score.groupby(by='platform')['platform'].count().reset_index(name='nan_count')
all_by_year = c_df.groupby(by='platform')['platform'].count().reset_index(name='all_count')
merged = pd.merge(nan_by_platform, all_by_year, on='platform')
merged['share_nan_in_all'] = round(merged['nan_count'] / merged['all_count'], 2)
merged

platform  nan_count  all_count  share_nan_in_all
0      2600        135        135              1.00
1       3DO          3          3              1.00
2       3DS        296        530              0.56
3        DC         38         52              0.73
4        DS        892       2177              0.41
5        GB         98         98              1.00
6       GBA        313        837              0.37
7        GC         89        563              0.16
8       GEN         29         29              1.00
9        GG          1          1              1.00
10      N64        323        323              1.00
11      NES        100        100              1.00
12       NG         12         12              1.00
13       PC        158        990              0.16
14     PCFX          1          1              1.00
15       PS       1005       1215              0.83
16      PS2        689       2189              0.31
17      PS3        376       1355              0.28
18      PS4        132        395              0.33
19      PSP        676       1229              0.55
20      PSV        282        435              0.65
21      SAT        174        174              1.00
22      SCD          6          6              1.00
23     SNES        241        241              1.00
24     TG16          2          2              1.00
25       WS          6          6              1.00
26      Wii        328       1340              0.24
27     WiiU         42        147              0.29
28     X360        214       1281              0.17
29       XB         89        839              0.11
30     XOne         54        251              0.22

In [27]:
# Строки с пустым 'rating'
nan_rating = c_df[c_df['rating'].isna()]
nan_rating.head()

name platform  year_of_release         genre  \
1          Super Mario Bros.      NES           1985.0      Platform   
4   Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
5                     Tetris       GB           1989.0        Puzzle   
9                  Duck Hunt      NES           1984.0       Shooter   
10                Nintendogs       DS           2005.0    Simulation   

    na_sales  eu_sales  jp_sales  other_sales  critic_score user_score rating  
1      29.08      3.58      6.81         0.77           NaN        NaN    NaN  
4      11.27      8.89     10.22         1.00           NaN        NaN    NaN  
5      23.20      2.26      4.22         0.58           NaN        NaN    NaN  
9      26.93      0.63      0.28         0.47           NaN        NaN    NaN  
10      9.05     10.95      1.93         2.74           NaN        NaN    NaN

In [28]:
# Проверяем зависимость rating от года выпуска игры
nan_by_year = nan_rating.groupby(by='year_of_release')['year_of_release'].count().reset_index(name='nan_count')
all_by_year = c_df.groupby(by='year_of_release')['year_of_release'].count().reset_index(name='all_count')
merged = pd.merge(nan_by_year, all_by_year, on='year_of_release')
merged['share_nan_in_all'] = round(merged['nan_count'] / merged['all_count'], 2)
merged

year_of_release  nan_count  all_count  share_nan_in_all
0            1980.0          9          9              1.00
1            1981.0         46         46              1.00
2            1982.0         37         37              1.00
3            1983.0         18         18              1.00
4            1984.0         14         14              1.00
5            1985.0         13         14              0.93
6            1986.0         22         22              1.00
7            1987.0         17         17              1.00
8            1988.0         14         15              0.93
9            1989.0         17         17              1.00
10           1990.0         16         16              1.00
11           1991.0         42         42              1.00
12           1992.0         41         43              0.95
13           1993.0         62         62              1.00
14           1994.0        120        121              0.99
15           1995.0        220        220              1.00
16           1996.0        260        267              0.97
17           1997.0        274        293              0.94
18           1998.0        352        384              0.92
19           1999.0        298        341              0.87
20           2000.0        207        357              0.58
21           2001.0        145        491              0.30
22           2002.0        177        839              0.21
23           2003.0        165        789              0.21
24           2004.0        167        771              0.22
25           2005.0        235        950              0.25
26           2006.0        336       1020              0.33
27           2007.0        381       1218              0.31
28           2008.0        452       1445              0.31
29           2009.0        423       1450              0.29
30           2010.0        465       1279              0.36
31           2011.0        439       1149              0.38
32           2012.0        310        670              0.46
33           2013.0        231        552              0.42
34           2014.0        237        584              0.41
35           2015.0        294        612              0.48
36           2016.0        224        507              0.44

In [29]:
# Проверяем зависимость rating от платформы игры
nan_by_platform = nan_rating.groupby(by='platform')['platform'].count().reset_index(name='nan_count')
all_by_year = c_df.groupby(by='platform')['platform'].count().reset_index(name='all_count')
merged = pd.merge(nan_by_platform, all_by_year, on='platform')
merged['share_nan_in_all'] = round(merged['nan_count'] / merged['all_count'], 2)
merged

platform  nan_count  all_count  share_nan_in_all
0      2600        135        135              1.00
1       3DO          3          3              1.00
2       3DS        298        530              0.56
3        DC         38         52              0.73
4        DS        892       2177              0.41
5        GB         98         98              1.00
6       GBA        308        837              0.37
7        GC         89        563              0.16
8       GEN         29         29              1.00
9        GG          1          1              1.00
10      N64        323        323              1.00
11      NES        100        100              1.00
12       NG         12         12              1.00
13       PC        203        990              0.21
14     PCFX          1          1              1.00
15       PS       1004       1215              0.83
16      PS2        689       2189              0.31
17      PS3        389       1355              0.29
18      PS4        139        395              0.35
19      PSP        676       1229              0.55
20      PSV        281        435              0.65
21      SAT        174        174              1.00
22      SCD          6          6              1.00
23     SNES        241        241              1.00
24     TG16          2          2              1.00
25       WS          6          6              1.00
26      Wii        326       1340              0.24
27     WiiU         42        147              0.29
28     X360        213       1281              0.17
29       XB         92        839              0.11
30     XOne         61        251              0.24

Промежуточные выводы:
 - 2 пропуска в столбце 'name' совпадают с 2 пустыми строками в 'genre', скорее всего это ошибка. Решение: из-за небольшого количества и неинформативности строки можно удалить.
 - 275 пропусков в столбце 'year_of_release' - никакой связи с жанром игры или платформой не нашел. Решение: в некоторых названиях игр существует год, например 'Madden NFL 2004', для таких игр можно использовать этот год как year_of_release (если год в пределах разумного), для остальных строк возьмем среднее значение года по платформе
 - 6 пропусков в столбце eu_sales -  можно заменить на среднее значение в зависимости от названия платформы и года выхода игры.
 - 4 пропусков в столбце jp_sales  можно заменить на среднее значение в зависимости от названия платформы и года выхода игры.
 - 8714 пропусков в столбце critic_score - до 1999 года почти  100% игр не имели critic_score, также существуют платформы (возможно они были популярны до как раз до 1999) все значения critic_score пустые. Абсолютно у всех платформ и годов существуют пропуски, значит они не ошибочны. Решение: можно ввести значение-индикатор, из df.describe знаем, что min = 13, max=98, т.е. значения варьируются от 0 до 100, значит можно значение-индиктор можно сделать -1.
 - 6804 пропусков в столбце user_score - все точно как и с critic_score
 - 6871 пропусков в столбце rating - все точно как и с critic_score, за исключением того, что заменяем пустые строки по следующему алгоритму - для каждого жанра игры находим долю каждого рейтинга и принимаем ее за вероятность появления ее в пустой строке

- Обработайте пропущенные значения. Для каждого случая вы можете выбрать оптимальный, на ваш взгляд, вариант: заменить на определённое значение, оставить как есть или удалить.
- Если вы решите заменить пропуски на значение-индикатор, то убедитесь, что предложенное значение не может быть использовано в данных.
- Если вы нашли пропуски в данных с количеством проданных копий игры в том или ином регионе, их можно заменить на среднее значение в зависимости от названия платформы и года выхода игры.

In [30]:
# Обработка 'name'
c_df = c_df.dropna(subset=['name'])

In [31]:
# Обработка 'year_of_release'
# Для игр где year_of_release = NaN, но в названии есть приписка о годе (например 'Madden NFL 2004'), подстваляем этот год в year_of_release

def check_year(row):
    try:
        year = int(row['name'][-4:])
        # Проверяем, что это год
        if year >= 1960 and year <= 2025:
            return year
        else:
            return row['year_of_release']
    except:
        return row['year_of_release']

c_df['year_of_release'] = c_df.apply(check_year, axis=1)

# Заполнение средним значением года по платформе для всех оставшихся year_of_release = NaN
def year_by_platform(row):
    if pd.isna(row['year_of_release']):
        group = c_df[c_df['platform'] == row['platform']]
        return round(group['year_of_release'].mean(), 0)
    else:
        return row['year_of_release']
    
c_df['year_of_release'] = c_df.apply(year_by_platform, axis=1)

# Меняем тип данных с float на int
c_df['year_of_release'] = c_df['year_of_release'].astype('int64')

In [32]:
# Обработка eu_sales

def eu_sales_filler(row):
    if pd.isna(row['eu_sales']):
        group = c_df[
            (c_df['platform'] == row['platform']) &
            (c_df['year_of_release'] == row['year_of_release']) &
            (c_df['genre'] == row['genre'])
        ]
        
        mean_sales = group['eu_sales'].mean()
        return mean_sales
    else:
        return row['eu_sales']
    
c_df['eu_sales'] = c_df.apply(eu_sales_filler, axis=1)

In [33]:
# Обработка jp_sales

def jp_sales_filler(row):
    if pd.isna(row['jp_sales']):
        group = c_df[
            (c_df['platform'] == row['platform']) &
            (c_df['year_of_release'] == row['year_of_release']) &
            (c_df['genre'] == row['genre'])
        ]
        
        mean_sales = group['jp_sales'].mean()
        return mean_sales
    else:
        return row['jp_sales']

c_df['jp_sales'] = c_df.apply(jp_sales_filler, axis=1)

In [34]:
# Обработка critic_score
c_df['critic_score'] = c_df['critic_score'].fillna(-1)

In [35]:
# Обработка user_score
c_df['user_score'] = c_df['user_score'].fillna(-1)

In [36]:
# Обработка rating

# Прежде чем применять алгоритм проврерим значения
print(c_df['rating'].unique()) # Найдено устаревшее значение 'K-A', в последствии его заменили на 'E', сделаем также
print()
print(c_df['genre'].unique()) # Видно, что значения повторяются в разных регистрах, нужно привести все к одному

['E' nan 'M' 'T' 'E10+' 'K-A' 'AO' 'EC' 'RP']

['Sports' 'Platform' 'Racing' 'Role-Playing' 'Puzzle' 'Misc' 'Shooter'
 'Simulation' 'Action' 'Fighting' 'Adventure' 'Strategy' 'MISC'
 'ROLE-PLAYING' 'RACING' 'ACTION' 'SHOOTER' 'FIGHTING' 'SPORTS' 'PLATFORM'
 'ADVENTURE' 'SIMULATION' 'PUZZLE' 'STRATEGY']


In [37]:
# Замена устаревшего рейтинга 'K-A' на 'E'
def rating_replace(row):
    if row['rating'] == 'K-A':
        return 'E'
    else:
        return row['rating']
    
c_df['rating'] = c_df.apply(rating_replace, axis=1)
    
    
# Привеодим 'genre' к нижнему регистру
c_df['genre'] = c_df['genre'].str.lower()


# Определние текущего распределения рейтинга для каждого жанра
rating_prob_by_genre = {}

for genre in c_df['genre'].unique():
    rating_prob_by_genre[genre] = c_df[c_df['genre']==genre].groupby(by='rating')['genre'].count()
    / c_df[c_df['genre']==genre].groupby(by='rating')['genre'].count().sum()

    
# Принимаем  распределние рейтинга по жанрам за вероятность того, что оно окажется в пустых ячейках и заполняем
import random

def get_rating(row):
    if pd.isna(row['rating']):
        cur_genre = row['genre']
        genre = rating_prob_by_genre[cur_genre].keys().to_list()
        weights = rating_prob_by_genre.get(cur_genre).to_list()
        res = random.choices(genre, weights=weights, k=1)
        return res[0]

    else:
        return row['rating']

c_df['rating'] = c_df.apply(get_rating, axis=1)

In [38]:
# Проврерка что все типы приведены к нужным и нет пропусков
c_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16954 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16954 non-null  object 
 2   year_of_release  16954 non-null  int64  
 3   genre            16954 non-null  object 
 4   na_sales         16954 non-null  float64
 5   eu_sales         16954 non-null  float64
 6   jp_sales         16954 non-null  float64
 7   other_sales      16954 non-null  float64
 8   critic_score     16954 non-null  float64
 9   user_score       16954 non-null  object 
 10  rating           16954 non-null  object 
dtypes: float64(5), int64(1), object(5)
memory usage: 1.6+ MB


### 2.4. Явные и неявные дубликаты в данных

- Изучите уникальные значения в категориальных данных, например с названиями жанра игры, платформы, рейтинга и года выпуска. Проверьте, встречаются ли среди данных неявные дубликаты, связанные с опечатками или разным способом написания.
- При необходимости проведите нормализацию данных с текстовыми значениями. Названия или жанры игр можно привести к нижнему регистру, а названия рейтинга — к верхнему.

In [39]:
# В платформах дубликатов нет
print(f"Уникальные значения в platform: \n{c_df['platform'].unique()} \n")
# На всякий случай, вдруг не заметил пробел
c_df['platform'] = c_df['platform'].str.rstrip()
c_df['platform'] = c_df['platform'].str.lstrip()


# В жанрах привели все к нижнему регистру тем самым убрав дубликаты в прудыдущем пункте
print(f"Уникальные значения в genre: \n{c_df['genre'].unique()}\n")

# Приводим name к нижнему регистру и убираем лишние пробелы справа и слева
c_df['name'] = c_df['name'].str.lower()
c_df['name'] = c_df['name'].str.rstrip()
c_df['name'] = c_df['name'].str.lstrip()

# В rating уже приведено к верхнему регистру и дубликатов нет
print(f"Уникальные значения в rating: \n{c_df['rating'].unique()}")

Уникальные значения в platform: 
['Wii' 'NES' 'GB' 'DS' 'X360' 'PS3' 'PS2' 'SNES' 'GBA' 'PS4' '3DS' 'N64'
 'PS' 'XB' 'PC' '2600' 'PSP' 'XOne' 'WiiU' 'GC' 'GEN' 'DC' 'PSV' 'SAT'
 'SCD' 'WS' 'NG' 'TG16' '3DO' 'GG' 'PCFX'] 

Уникальные значения в genre: 
['sports' 'platform' 'racing' 'role-playing' 'puzzle' 'misc' 'shooter'
 'simulation' 'action' 'fighting' 'adventure' 'strategy']

Уникальные значения в rating: 
['E' 'T' 'M' 'E10+' 'AO' 'EC' 'RP']


- После того как нормализуете данные и устраните неявные дубликаты, проверьте наличие явных дубликатов в данных.

In [40]:
print(f"Количество явных дубликатов в данных: {c_df.duplicated().sum()}")

c_df = c_df.drop_duplicates(subset=None, keep='first', inplace=False)
print(f"Количество оставшихся строк: {c_df.shape[0]}")

Количество явных дубликатов в данных: 176
Количество оставшихся строк: 16778


In [41]:
# Считаем, что такие игры тоже считаются дубликатами и удаляем их

temp = c_df.groupby(by=['name', 'platform', 'year_of_release', 'genre'], as_index=True)['genre'].count()
print(f"Количество игр с одинаковым названием, платформой, годом релиза, и жанром: {len(temp[temp > 1])}")
temp[temp > 1].head()

Количество игр с одинаковым названием, платформой, годом релиза, и жанром: 66


name                                       platform  year_of_release  genre       
999: nine hours, nine persons, nine doors  DS        2009             adventure       2
beyblade burst                             3DS       2016             role-playing    2
biohazard: revival selection               PS3       2011             action          2
breath of fire iii                         PS        1997             role-playing    2
broken helix                               PS        1997             action          2
Name: genre, dtype: int64

In [42]:
# Удаляем дубликаты с одинаковым названием, платформой, годом релиза, и жанром
c_df = c_df.drop_duplicates(subset=['name', 'platform', 'year_of_release', 'genre'], keep='first', inplace=False)

In [51]:
c_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16712 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16712 non-null  object 
 1   platform         16712 non-null  object 
 2   year_of_release  16712 non-null  int64  
 3   genre            16712 non-null  object 
 4   na_sales         16712 non-null  float64
 5   eu_sales         16712 non-null  float64
 6   jp_sales         16712 non-null  float64
 7   other_sales      16712 non-null  float64
 8   critic_score     16712 non-null  float64
 9   user_score       16712 non-null  object 
 10  rating           16712 non-null  object 
dtypes: float64(5), int64(1), object(5)
memory usage: 1.5+ MB


- Напишите промежуточный вывод: укажите количество найденных дубликатов и действия по их обработке.

- В процессе подготовки данных вы могли что-либо удалять, например строки с пропусками или ошибками, дубликаты и прочее. В этом случае посчитайте количество удалённых строк в абсолютном и относительном значениях.

- После проведения предобработки данных напишите общий промежуточный вывод.

---

## 3. Фильтрация данных

Коллеги хотят изучить историю продаж игр в начале XXI века, и их интересует период с 2000 по 2013 год включительно. Отберите данные по этому показателю. Сохраните новый срез данных в отдельном датафрейме, например `df_actual`.

---

## 4. Категоризация данных
    
Проведите категоризацию данных:
- Разделите все игры по оценкам пользователей и выделите такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).

- Разделите все игры по оценкам критиков и выделите такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).

- После категоризации данных проверьте результат: сгруппируйте данные по выделенным категориям и посчитайте количество игр в каждой категории.

- Выделите топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

---

## 5. Итоговый вывод

В конце напишите основной вывод и отразите, какую работу проделали. Не забудьте указать описание среза данных и новых полей, которые добавили в исходный датасет.